### ISSUE TITLE
percolator reported mass seem to be wrong? (input from comet)

### ISSUE THREAD

**Issue by infusini** [crux-toolkit/crux-toolkit#453]
Hi,

I can't figure out the mass values reported by "percolator".

The input MGF is this:
```
TITLE=RawFile: P3141_HC67_1_01_9191 Index: 29964 precursor: 2 Charge: 5 
INSTRUMENT=ESI_QUAD_TOF
PEPMASS=413.43444905 1.442980e+05
CHARGE=5+
RTINSECONDS=2702.946726
SCANS=452
```
the Experimental neutral mass of the ion is:
`round( (413.43444905 * 5 ) - (5 * 1.0072765), 6)`   --> 2062.135863 which is exactly the mass reported in the `ExpMass` column by Comet output pin file (see the line below), and the `Mass` column seems to be the correct experimental M+H mass. The `CalcMass` as well seems correct neutral mass calculated from the peptide sequence.

the `pin` output from `crux comet` is:
```
SpecId	Label	ScanNr	ExpMass	CalcMass	Mass	dM	absdM	Peptide
P3141_HC67_1_01_9191_xic.comet_452_5_1	1	452	2062.135863	2062.143592	2063.14314	-4.000000E-006	4.000000E-006	K.NAVVQGLEQPHGLVVHPLR.G
```

In the percolator.target.psm.txt the `peptide mass` column seems to be the truncated version of the `.pin` column `CalcMass`:
```
file_idx	scan	charge	spectrum precursor m/z	spectrum neutral mass	peptide mass	sequence
306	452	5	413.233	2061.12860000	2062.14350000	NAVVQGLEQPHGLVVHPLR
```
the `spectrum precursor m/z` seems to be the correct calculation of the m/z mass of a 5+ ion using the value from `spectrum neutral mass` column.

So I can't figure out where the `spectrum neutral mass` value comes from, as it shouldn't be less than the calculated mass. Is there anything I'm missing?

Thanks!

**Comment by wfondrie** [comment]
Hi @infusini,

That does seem odd. I'm looking into this now and trying to reproduce it with a different file. I have a couple questions:
1.  Do you see the odd `spectrum neutral mass` for this PSM in the `comet.target.txt` file as well? 
2. What parameters did you use with to run `crux comet`?

**Comment by infusini** [comment]
Hi @wfondrie,

I don't have the `comet.target.txt` files, I have been writing out just the .pin files. But if needed I can run a separate search.

I put the parameters file in this [gist](https://gist.github.com/infusini/4a2ae58be9a623c7eccc334885f9a05f)

Cheers.

**Comment by wfondrie** [comment]
Thank you for providing those parameters.

I've been able to reproduce the odd `spectrum neutral mass` and `spectrum precursor m/z` you saw in your crux percolator results. I verified that this is introduced after Percolator as well. 

For reference, here is an example of a PSM from my result files:

- In comet.target.txt, `spectrum neutral mass` = `1228.53411`
- In comet.target.pin, `ExpMass` = `1228.53411`
- In percolator.target.psms.txt, `spectrum neutral mass` = `1227.5268`

Unfortunately, this appears to be due to how Comet writes the pin file for Percolator---crux expects the `ExpMass` to be the singly charged mass, not the neutral mass of the, spectrum. (Details below). The good news is that that this should have no effect on your results, other than providing the incorrect `spectrum neutral mass` and `spectrum precursor m/z` in the percolator output. 

---
### Details
The adapter for Percolator expects the `ExpMass` column to be the singly charged mass:
https://github.com/crux-toolkit/crux-toolkit/blob/98ef587252c12c05e8c2816148e33185521fbb16/src/app/PercolatorAdapter.cpp#L121-L126

However, Comet writes `ExpMass` as the neutral mass instead (lines 172-173 in [CometWritePercolator.cpp](https://sourceforge.net/p/comet-ms/code/HEAD/tree/trunk/comet-ms/CometSearch/CometWritePercolator.cpp#l172)):
```c++
      fprintf(fpout, "%0.6f\t", pQuery->_pepMassInfo.dExpPepMass - PROTON_MASS);  //ExpMass
      fprintf(fpout, "%0.6f\t", pOutput[iWhichResult].dPepMass - PROTON_MASS);  //CalcMass
```
It looks like to fix this, we need a formal definition of what the `ExpMass` column should be and enforce it for Tide-search and Comet. What do you think @wsnoble?





`

**Comment by infusini** [comment]
No worries @wfondrie. You are right, the issue does not affect the results, in fact I noticed it just by chance while trying to find a better way to merge the result table to my own feature table
But thanks for checking.

**Comment by wsnoble** [comment]
Thanks for tracking this down, Will. I agree that we just need to make a
decision here and then either change tide-search or comet output. Lukas,
can you comment on whether ExpMass in Percolator is supposed to be the
singly charged mass or the neutral mass of the spectrum?

**Comment by wsnoble** [comment]
Happy to make this fix in Comet.

**Comment by infusini** [comment]
Hi, we found another inconsistency between comet and percolator outputs.

Specifically, in `C` containing peptides percolator reports the mass of the peptide sequence only, without considering that in the `params` file we specify the `+57.` of alkylation, while comet reports the "correct" the peptide + alkylation mass.

```
# Specify a static modification to the residue C.
# Available for comet.
add_C_cysteine=57.021464
```

**Comment by wsnoble** [comment]
OK, will do.

**Comment by wsnoble** [comment]
Yes, let's please go ahead and make this change in Comet.  Thanks, Jimmy!

**Comment by jvolkening** [comment]
Hello,

I've noticed that this issue still seems to persist (v4.1, installed today) in the tab-delimited output created by `crux percolator`. Issue is just as described above -- calculated peptide masses for C-containing peptides are off by the mass of the carbamidomethylation. Possibly happens for other fixed mods as well, I haven't tested.

Values in the pepXML output look good. Is it possible this was fixed for the one output and not the other?

Thanks.

**Comment by jke000** [comment]
I'm able to confirm that Comet will output the correct mass for peptides containing the +57 static mod on cysteines.  But after running Crux Percolator, the "spectrum neutral mass" (column 5) contains the right mass but "peptide mass" (column 6) is missing the +57 in the "percolator.target.psms.txt".  I'm guessing that Percolator is calculating the "peptide mass" column and there's no mechanism in the .pin input to notify Percolator that there are static modifications in the identified peptides.  Here's one example peptide from the Comet generated .pin output:  K.HTGPGILSM[15.9949]ANAGPNTNGSQFFICTAK.T.  To address this particular issue, presumably I should make Comet write the peptide output as K.HTGPGILSM[15.9949]ANAGPNTNGSQFFIC[57.0215]TAK.T and add in any other static modification masses within these peptide strings.  Can the Crux team confirm that this would be the correct peptide string format/convention for .pin files?  Presumably only Percolator consumes the .pin files but would this change possibly break anything else?

**Comment by wsnoble** [comment]
I don't like the idea of Comet having to explicitly write out the C+57 modification. Typically that's implicit. We should make sure Percolator is smart enough to add the mod on its own.  However, I need to first figure out if Percolator is calculating the "peptide mass" value internally or if this is happening in the Crux part of the code.

**Comment by wsnoble** [comment]
FYI, Charles says that we are currently pulling commit 3c62af2 from the Comet GitHub, which is from Dec. 9, 2021. We add a CMakeLists.txt file, but don't make any other patches to the Comet source.

**Comment by wsnoble** [comment]
It appears that this problem with the peptide mass is within the Crux portion of the code.
